tests/c/tcp-push-pop.c

```diff
@@ -0,0 +1,166 @@
+    assert(demi_sgafree(&res.qr_value.sga) == 0);
+    return recv_bytes;
+}
+
+static void run_server(struct sockaddr_in *addr)
+{
+    unsigned int recv_bytes = 0;
+    int sockqd = accept_get_newsockqd(addr);
+
+    while (recv_bytes < MAX_BYTES)
+    {
+        recv_bytes += pop_get_received_nbytes(sockqd);
+        fprintf(stdout, "pop: total bytes received: (%d)\n", recv_bytes);
+    }
+
+    assert(recv_bytes == MAX_BYTES);
+}
+
+static int connect_get_sockqd(const struct sockaddr_in *const addr)
+{
+    int sockqd = -1;
+    demi_qtoken_t tok = -1;
+    demi_qresult_t res = {0};
+
+    assert(demi_socket(&sockqd, AF_INET, SOCK_STREAM, 0) == 0);
+    assert(demi_connect(&tok, sockqd, (const struct sockaddr *)addr, sizeof(struct sockaddr_in)) == 0);
+    assert(demi_wait(&res, tok) == 0);
+    assert(res.qr_opcode == DEMI_OPC_CONNECT);
+    return sockqd;
+}
+
+static int push_get_sent_nbytes(const struct sockaddr_in *const addr, const int sockqd)
+{
+    demi_qtoken_t tok = -1;
+    demi_qresult_t res = {0};
+    demi_sgarray_t sga = demi_sgaalloc(DATA_SIZE);
+    int sent_bytes = 0;
+
+    assert(sga.sga_segs != 0);
+    memset(sga.sga_segs[0].sgaseg_buf, 1, DATA_SIZE);
+    assert(demi_pushto(&tok, sockqd, &sga, (const struct sockaddr *)addr, sizeof(struct sockaddr_in)) == 0);
+    assert(demi_wait(&res, tok) == 0);
+    assert(res.qr_opcode == DEMI_OPC_PUSH);
+    sent_bytes = sga.sga_segs[0].sgaseg_len;
+    assert(demi_sgafree(&sga) == 0);
+    return sent_bytes;
+}
+
+static void run_client(const struct sockaddr_in *const addr)
+{
+    unsigned int sent_bytes = 0;
+    int sockqd = connect_get_sockqd(addr);
+
+    while (sent_bytes < MAX_BYTES)
+    {
+        sent_bytes += push_get_sent_nbytes(addr, sockqd);
+        fprintf(stdout, "push: total bytes sent: (%d)\n", sent_bytes);
+    }
+
+    assert(sent_bytes == MAX_BYTES);
+}
+
+static void usage(const char *const progname)
+{
+    fprintf(stderr, "Usage: %s MODE local-ipv4 local-port [remote-ipv4] [remote-port]\n", progname);
+    fprintf(stderr, "Modes:\n");
+    fprintf(stderr, "  --client    Launch program in client mode.\n");
+    fprintf(stderr, "  --server    Launch program in server mode.\n");
+}
+
+void build_addr(const char *const ip_str, const char *const port_str, struct sockaddr_in *const addr)
+{
+    int port = -1;
+
+    sscanf(port_str, "%d", &port);
+    addr->sin_family = AF_INET;
+    addr->sin_port = htons(port);
+    assert(inet_pton(AF_INET, ip_str, &addr->sin_addr) == 1);
+}
+
+// Exercises a one-way direction communication through TCP. This system-level test instantiates two demikernel peers: a
+// client and a server. The client sends TCP packets to the server in a tight loop. The server process is a tight loop
+// received TCP packets from the client.
+int main(int argc, const char *argv[])
+{
+    struct sockaddr_in addr = {0};
+
+    reg_sighandlers();
+    assert(demi_init(argc, argv) == 0); // needed by both server and client
```

Comment:
Ok, we current get a compilation warning in this line because my fix on the signature of `demi_init()` was not merged in yet.